trin-core/src/types/block_body.rs

```diff
@@ -0,0 +1,307 @@
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use bytes::Bytes;
+use eth_trie::{EthTrie, MemoryDB, Trie};
+use ethereum_types::{Address, H160, H256, U256, U64};
+use rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream};
+use rlp_derive::{RlpDecodable, RlpEncodable};
+use sha3::{Digest, Keccak256};
+
+use super::{header::Header, receipts::TransactionId};
+
+#[derive(Clone, Debug)]
+pub struct BlockBody {
+    _txs: Transactions,
+    _uncle_list: Vec<Header>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Transactions {
+    tx_list: Vec<Transaction>,
+}
+
+impl Transactions {
+    pub fn root(&self) -> anyhow::Result<H256> {
+        let memdb = Arc::new(MemoryDB::new(true));
+        let mut trie = EthTrie::new(Arc::clone(&memdb));
+
+        // Insert txs into tx tree
+        for (index, tx) in self.tx_list.iter().enumerate() {
+            let path = rlp::encode(&index).freeze().to_vec();
+            let encoded_tx = tx.encode();
+            trie.insert(&path, &encoded_tx)
+                .map_err(|err| anyhow!("Error calculating transactions root: {err:?}"))?;
+        }
+
+        trie.root_hash()
+            .map_err(|err| anyhow!("Error calculating transactions root: {err:?}"))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Uncles {
+    uncle_list: Vec<Header>,
+}
+
+impl Uncles {
+    pub fn root(&self) -> anyhow::Result<H256> {
+        // generate rlp encoded list of uncles
+        let mut stream = RlpStream::new();
+        stream.append_list(&self.uncle_list);
+        let uncles_rlp = stream.out().freeze();
+
+        // hash rlp uncles
+        let mut keccak = Keccak256::new();
+        keccak.update(uncles_rlp);
+        let hash = keccak.finalize().to_vec();
+
+        Ok(H256::from_slice(&hash))
+    }
+}
+
+#[derive(Eq, Debug, Clone, PartialEq)]
+pub enum Transaction {
+    Legacy(LegacyTransaction),
+    AccessList(AccessListTransaction),
+    EIP1559(EIP1559Transaction),
+}
+
+impl Transaction {
+    pub fn decode(transaction: &[u8]) -> Result<Self, DecoderError> {
+        if transaction.is_empty() {
+            // at least one byte needs to be present
+            return Err(DecoderError::RlpIncorrectListLen);
+        }
+        let id = TransactionId::try_from(transaction[0]);
+        if id.is_err() {
+            return Err(DecoderError::Custom("Unknown transaction type"));
```

Comment:
Maybe an `if let` so we can format the unknown byte into this error message.

Reply:
`DecoderError::Custom()` only accepts a `str` and not a `String` - making it tricky to include the error message in there since the variable doesn't live long enough

Reply:
Oh I've run into that with a couple other APIs too, there must be some better solution that we're missing...